app/models/concerns/user_omniauth_concern.rb

```diff
@@ -0,0 +1,52 @@
+module UserOmniauthConcern
+  extend ActiveSupport::Concern
+
+  included do
+    devise :omniauthable, omniauth_providers: [:facebook]
+  end
+
+  module ClassMethods
+    # Find or create a user from omniauth data.
+    #
+    # @param [Hash] auth The data with omniauth info.
+    # @return [User] The user found or created from omniauth data.
+    def from_omniauth(auth)
+      identity = User::Identity.find_by(provider: auth.provider, uid: auth.uid)
+      return identity.user if identity
+
+      User.create do |user|
+        user.update_from_general_auth_data(auth)
+        user.instance_users.build
+      end
+    end
+
+    # A convenience method that receives both parameters and session to initialize a user.
+    # This method is a override of devise {#new_with_session}
+    def new_with_session(params, session)
+      super.tap do |user|
+        facebook_data = session['devise.facebook_data']
```

Comment:
@lowjoel The data is used here.

And the data starting with devise was because:
>In case the user is not persisted, we store the OmniAuth data in the session. Notice we store this data using "devise." as key namespace. This is useful because Devise removes all the data starting with "devise." from the session whenever a user signs in, so we get automatic session clean up.

REF: https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview#facebook-example